frontends/PyCDE/src/pycde/matrix.py

```diff
@@ -0,0 +1,264 @@
+    # not passed to numpy. This dimension would instead refer to an access
+    # into the inner data type.
+    # This access would allow for
+    # - bitwise access for integer data types
+    # - struct access for structs
+    # This issue, however, is a more general one, since we don't currently
+    # support lhs assignment of PyCDE Value's, which would be require for this.
+
+    # Infer the target shape based on the access to the numpy array.
+    # circt_to_arr will then try to convert the value to this shape.
+    v = self.circt_to_arr(value, self.target_shape_for_idxs(np_access))
+    super().__setitem__(np_access, v)
+
+  def check_is_fully_assigned(self):
+    """ Checks that all sub-matrices have been fully assigned. """
+    unassigned = np.argwhere(self == None)
+    if len(unassigned) > 0:
+      raise ValueError(f"Unassigned sub-matrices: \n{unassigned}")
+
+  def assign_default_driver(self, value):
+    """Assigns a default driver to any unassigned value in the matrix"""
+    for arg in np.argwhere(self == None):
+      super().__setitem__(tuple(arg), value)
+
+  def to_circt(self, create_wire=True, dtype=None, default_driver=None):
+    """Materializes this matrix to a ListValue through hw.array_create operations.
+    
+    if 'create_wire' is True, the matrix will be materialized to an sv.wire operation
+    and the returned value will be a read-only reference to the wire.
+    This wire acts as a barrier in CIRCT to prevent dataflow optimizations
+    from reordering/optimizing the materialization of the matrix, which might
+    reduce debugability.
+
+    If default_driver is set, any unassigned value will be assigned to this. The
+    default driver is expected to be of equal type as the dtype of this matrix.
+
+    """
+    if self.circt_output:
+      return self.circt_output
+
+    if dtype == None:
+      dtype = self.pycde_dtype
+
+    if default_driver:
+      if default_driver.type != self.pycde_dtype:
+        raise ValueError(
+            f"Default driver {default_driver} is not of type {dtype}.")
+      self.assign_default_driver(default_driver)
+
+    # Check that the entire matrix has been assigned. If not, an exception is
+    # thrown.
+    self.check_is_fully_assigned()
+
+    def build_subarray(lstOrVal):
+      # Recursively converts this matrix into ListValues through hw.array_create
+      # operations.
+      if not isinstance(lstOrVal, BitVectorValue):
+        subarrays = [build_subarray(v) for v in lstOrVal]
+        return hw.ArrayCreateOp(subarrays)
+      return lstOrVal
+
+    # Materialize the matrix
+    self.circt_output = build_subarray(self)
+
+    if create_wire:
+      wire = sv.WireOp(self.circt_output.type, self.name + "_wire")
+      sv.AssignOp(wire, self.circt_output)
+      self.circt_output = wire.read
+
+    return self.circt_output
+
+
+"""
+Inject a curated list of numpy functions into the ListValue class.
```

Comment:
Please just add these to ListValue directly with `def`s for each one. When debugging, we should make it clear to see what's going on.